zuul-core/src/main/java/com/netflix/zuul/netty/server/psk/ZuulPskServer.java

```diff
@@ -0,0 +1,230 @@
+
+    private static class PSKTimings {
+        private final Timer handshakeCompleteTimer;
+
+        private Long handshakeStartTime;
+
+        PSKTimings(Registry registry) {
+            handshakeCompleteTimer = registry.timer("zuul.psk.handshake.complete.time");
+        }
+
+        public void recordHandshakeStarting() {
+            handshakeStartTime = System.nanoTime();
+        }
+
+        public void recordHandshakeComplete() {
+            handshakeCompleteTimer.record(System.nanoTime() - handshakeStartTime, TimeUnit.NANOSECONDS);
+        }
+    }
+
+    private final PSKTimings pskTimings;
+
+    private final ExternalTlsPskProvider externalTlsPskProvider;
+
+    private final ChannelHandlerContext ctx;
+
+    private final Set<ProtocolName> supportedApplicationProtocols;
+
+    public ZuulPskServer(
+            TlsCrypto crypto,
+            Registry registry,
+            ExternalTlsPskProvider externalTlsPskProvider,
+            ChannelHandlerContext ctx,
+            Set<ProtocolName> supportedApplicationProtocols) {
+        super(crypto);
+        this.pskTimings = new PSKTimings(registry);
+        this.externalTlsPskProvider = externalTlsPskProvider;
+        this.ctx = ctx;
+        this.supportedApplicationProtocols = supportedApplicationProtocols;
+    }
+
+    @Override
+    public TlsCredentials getCredentials() {
+        return null;
+    }
+
+    @Override
+    protected Vector getProtocolNames() {
+        Vector protocolNames = new Vector();
+        if (supportedApplicationProtocols!=null) {
+            supportedApplicationProtocols.forEach(protocolNames::addElement);
+        }
+        return protocolNames;
+    }
+
+    @Override
+    public void notifyHandshakeBeginning() throws IOException {
+        pskTimings.recordHandshakeStarting();
+        this.ctx.channel().attr(TLS_HANDSHAKE_USING_EXTERNAL_PSK).set(false);
+        // TODO: sunnys - handshake timeouts
+        super.notifyHandshakeBeginning();
+    }
+
+    @Override
+    public void notifyHandshakeComplete() throws IOException {
+        pskTimings.recordHandshakeComplete();
+        this.ctx.channel().attr(TLS_HANDSHAKE_USING_EXTERNAL_PSK).set(true);
+        super.notifyHandshakeComplete();
+        ctx.fireUserEventTriggered(SslHandshakeCompletionEvent.SUCCESS);
+    }
+
+    @Override
+    protected ProtocolVersion[] getSupportedVersions() {
+        return ProtocolVersion.TLSv13.only();
+    }
+
+    @Override
+    protected int[] getSupportedCipherSuites() {
+        return TlsUtils.getSupportedCipherSuites(
+                getCrypto(),
+                TlsPskHandler.SUPPORTED_TLS_PSK_CIPHER_SUITE_MAP.keySet().stream()
+                        .mapToInt(Number::intValue)
+                        .toArray());
+    }
+
+    @Override
+    public ProtocolVersion getServerVersion() throws IOException {
+        return super.getServerVersion();
+    }
+
+    @Override
+    @SneakyThrows
```

Comment:
Are there other checked exceptions that you're not handling?

Reply:
Sunny asked the BC folks but we havent received the response yet
https://github.com/bcgit/bc-java/issues/1673

Reply:
Todo Deepti : Why we are using @SneakyThrows